includes/actions/class-amp-canonical-mode-actions.php

```diff
@@ -0,0 +1,220 @@
+<?php
+/**
+ * Class AMP_Canonical_Mode_Actions
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Canonical_Mode_Actions
+ *
+ * Callbacks for adding AMP-related things to the theme when in canonical mode.
+ */
+class AMP_Canonical_Mode_Actions {
+
+	const COMPONENT_SCRIPTS_PLACEHOLDER = '<!--AMP_COMPONENT_SCRIPTS_PLACEHOLDER-->';
+
+	/**
+	 * Register hooks.
+	 */
+	public static function register_hooks() {
+
+		// Remove core actions which are invalid AMP.
+		remove_action( 'wp_head', 'locale_stylesheet' );
+		remove_action( 'wp_head', 'print_emoji_detection_script', 7 );
+		remove_action( 'wp_head', 'wp_print_styles', 8 );
+		remove_action( 'wp_head', 'wp_print_head_scripts', 9 );
+		remove_action( 'wp_head', 'wp_custom_css_cb', 101 );
+		remove_action( 'wp_footer', 'wp_print_footer_scripts', 20 );
+		remove_action( 'wp_print_styles', 'print_emoji_styles' );
+
+		// Replace core's canonical link functionality with one that outputs links for non-singular queries as well. See WP Core #18660.
+		remove_action( 'wp_head', 'rel_canonical' );
+		add_action( 'wp_head', array( __CLASS__, 'rel_canonical' ), 1 );
+
+		// Add additional markup required by AMP <https://www.ampproject.org/docs/reference/spec#required-markup>.
+		add_action( 'wp_head', array( __CLASS__, 'print_meta_charset' ), 0 );
+		add_action( 'wp_head', array( __CLASS__, 'print_meta_viewport' ), 2 );
+		add_action( 'wp_head', array( __CLASS__, 'print_amp_boilerplate_code' ), 3 );
+		add_action( 'wp_head', array( __CLASS__, 'print_amp_scripts' ), 4 );
+		add_action( 'wp_head', array( __CLASS__, 'print_amp_custom_style' ), 5 );
```

Comment:
Thinks like `add_schemaorg_metadata()`, `add_analytics_data()` or `add_generator_metadata()` added in paired mode are not added in canonical mode, is that intentional?

Reply:
Just not implemented yet.